### PR TITLE
v.hull: Dereference after Null check

### PR DESCRIPTION
**Pull request by ShubhamDesai**
This pull request addresses the issue identified by coverity scan (CID: 1207412)

- if (edges) is added before e = edges->next; and entering the loop to ensure that it is not NULL before accessing next member.
- before accessing the member e->delete check whether e is not NULL.
- while (e != edges && e != NULL); here added check e!=NULL to ensure that loop terminates as soon as e is NULL.

### CODE REVIEW

File: vector/v.hull/chull.c [hunk 1]
```diff
@@ -687,16 +687,18 @@ void CleanEdges(void)
...
+        e = edges->next;
+        do {
+            if (e && e->delete) {
+                t = e;
+                e = e->next;
+                DELETE(edges, t);
+            }
+            else
+                e = e->next;
+        } while (e != edges && e != NULL);
```
Comment (metzm) [inline vector/v.hull/chull.c:700]:
```suggestion
        } while (e != edges);
```
Same as above.

**Comment by ShubhamDesai** [comment]
> If my understanding of the above is right, it should probably be enough to add a `if (!edges) return;` before `e = edges->next;`.

i guess using if (edges) will work the same 
because for !edges using return doesn't look good because function is void it may lead to additional warning in cppcheck i think.

File: vector/v.hull/chull.c [hunk 2]
```diff
@@ -687,16 +687,18 @@ void CleanEdges(void)
...
-            e = e->next;
-            DELETE(edges, t);
-        }
-        else
-            e = e->next;
-    } while (e != edges);
+    if (edges) {
+        e = edges->next;
+        do {
+            if (e && e->delete) {
```
[the remark this answers is not in the document]
Comment (ShubhamDesai) [reply vector/v.hull/chull.c]: done

File: vector/v.hull/chull.c (hunk 1 above)
Comment (ShubhamDesai) [reply vector/v.hull/chull.c]: done

**Comment by nilason** [comment]
> > If my understanding of the above is right, it should probably be enough to add a `if (!edges) return;` before `e = edges->next;`.
> 
> i guess using if (edges) will work the same because for !edges using return doesn't look good because function is void it may lead to additional warning in cppcheck i think.

In effect they are the same. `return;` is perfectly valid and only way to return early in a void function (note: there is no return _value_). Let's compare them side by side:

```c
    if (edges) {
        e = edges->next;
        do {
            if (e->delete) {
                t = e;
                e = e->next;
                DELETE(edges, t);
            }
            else
                e = e->next;
        } while (e != edges);
    }
```

```c
    if (!edges)
        return;

    e = edges->next;
    do {
        if (e->delete) {
            t = e;
            e = e->next;
            DELETE(edges, t);
        }
        else
            e = e->next;
    } while (e != edges);
```

In my opinion the readability and intent of the code is much better with the latter (tip: search for "_early return_" pattern and "_pyramid of doom_" anti-pattern).

To the root cause of this issue. I managed to reproduce this with Clang's `scan-build`[^1], and it became more clear whence it originates. The macro `DELETE` may set `edges` to NULL if `edges == edges->next`:

https://github.com/OSGeo/grass/blob/241589424d9612b913ef7f5502839ee6ab2151bc/vector/v.hull/macros.h#L40-L43

We have the same problem with CleanFaces() and a similar one with CleanVertices() too.

@metzm I'm not quite sure of the implications if `edges == edges->next` and we exit the function early. Was everything that were supposed to be cleaned up, actually cleaned up to that point?


[^1]:  The analysis report is available here as a html bundle: [scan-build-2024-10-16-105627-63060-1.zip](https://github.com/user-attachments/files/17393244/scan-build-2024-10-16-105627-63060-1.zip)

**Review by metzm — APPROVED** [review]

**Comment by metzm** [comment]
> We have the same problem with CleanFaces() and a similar one with CleanVertices() too.
indeed
> 
> @metzm I'm not quite sure of the implications if `edges == edges->next` and we exit the function early. Was everything that were supposed to be cleaned up, actually cleaned up to that point?

I think exiting the function early with
```
if (!edges)
        return;
```
is safe because there would be no edges left. `edges` is the root pointer to the double linked list, if it is NULL, the list is empty.

**Comment by echoix** [comment]
@nilason Or others, if you are ok with Markus Metz's response and approval, let's merge this. On Monday I'll get it merged if everything is still fine

**Review by nilason — CHANGES_REQUESTED** [review]
Let us address CID: 1207410 and CID: 1207411 too, with this PR.

@metzm Thank you very much for your input!

**Comment by ShubhamDesai** [comment]
> Let us address CID: 1207410 and CID: 1207411 too, with this PR.
> 
> @metzm Thank you very much for your input!

I did the changes.
also as you suggested @nilason i did the early return. I agree with you to exit early

**Review by nilason — APPROVED** [review]
@ShubhamDesai  Thanks!

**Pull request merged**